google/cloud/internal/invoke_result_test.cc

```diff
@@ -0,0 +1,88 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/internal/invoke_result.h"
+#include <gmock/gmock.h>
+
+namespace google {
+namespace cloud {
+inline namespace GOOGLE_CLOUD_CPP_NS {
+namespace internal {
+namespace {
+TEST(ResultOfTest, Lambda) {
+  // This test mostly uses static assertions because the result_of classes are
+  // largely meta-functions.
+
+  auto l = [](long, int) -> int { return 7; };
+
+  using F = decltype(l);
+
+  using R = invoke_result<F, long, int>::type;
+
+  static_assert(std::is_same<R, int>::value,
+                "expected `int` in invoke_result_t<>");
+}
+
+std::string test_function(int, std::string const&) { return "42"; }
+
+TEST(ResultOfTest, Function) {
+  // This test mostly uses static assertions because the result_of classes are
+  // largely meta-functions.
+
+  using R1 = decltype(invoker_function(test_function, 7, std::string{}));
+  static_assert(std::is_same<R1, std::string>::value,
+                "expected `std::string` in decltype(invoker_function())");
+
+  using R2 = decltype(invoker_function(test_function, std::declval<int>(),
+                                       std::declval<std::string const&>()));
+  static_assert(std::is_same<R2, std::string>::value,
+                "expected `std::string` in decltype(invoker_function())");
```

Comment:
Perhaps add R1 / R2 etc in the static_assert message so that it is easier to distinguish which one failed.

Reply:
Done. Thanks!